service/http/topic.go

```diff
@@ -301,9 +325,128 @@ func (s *Server) AddTopicEventHandler(sub *common.Subscription, fn common.TopicE
+	}
+	if err := s.topicRegistrar.AddBulkSubscription(sub, fn, maxMessagesCount, maxAwaitDurationMs); err != nil {
+		return err
+	}
+
+	s.mux.Handle(sub.Route, optionsHandler(http.HandlerFunc(
+		func(w http.ResponseWriter, r *http.Request) {
+			// check for post with no data
+			var (
+				body []byte
+				err  error
+			)
+			if r.Body != nil {
+				body, err = io.ReadAll(r.Body)
+				if err != nil {
+					http.Error(w, err.Error(), PubSubHandlerDropStatusCode)
+					return
+				}
+			}
+			if len(body) == 0 {
+				http.Error(w, "nil content", PubSubHandlerDropStatusCode)
+				return
+			}
+
+			// deserialize the event
+			var ins internal.BulkSubscribeEnvelope
+			if err = json.Unmarshal(body, &ins); err != nil {
+				http.Error(w, err.Error(), PubSubHandlerDropStatusCode)
+				return
+			}
+
+			statuses := make([]BulkSubscribeResponseEntry, 0, len(ins.Entries))
+
+			for _, entry := range ins.Entries {
+				itemJSON, err := json.Marshal(entry.Event)
+				if err != nil {
+					http.Error(w, err.Error(), PubSubHandlerDropStatusCode)
+					return
+				}
+				var in topicEventJSON
+
+				if err := json.Unmarshal(itemJSON, &in); err != nil {
+					http.Error(w, err.Error(), PubSubHandlerDropStatusCode)
+					return
+				}
+				if in.PubsubName == "" {
+					in.Topic = sub.PubsubName
+				}
+				if in.Topic == "" {
+					in.Topic = sub.Topic
+				}
+				data, rawData := in.getData()
+
+				te := common.TopicEvent{
+					ID:              in.ID,
+					SpecVersion:     in.SpecVersion,
+					Type:            in.Type,
+					Source:          in.Source,
+					DataContentType: in.DataContentType,
+					Data:            data,
+					RawData:         rawData,
+					DataBase64:      in.DataBase64,
+					Subject:         in.Subject,
+					PubsubName:      in.PubsubName,
+					Topic:           in.Topic,
+				}
+
+				retry, err := fn(r.Context(), &te)
+				if err == nil {
+					statuses = append(statuses, BulkSubscribeResponseEntry{
+						EntryId: entry.EntryId,
+						Status:  Success,
+					},
+					)
+				} else if retry {
+					statuses = append(statuses, BulkSubscribeResponseEntry{
+						EntryId: entry.EntryId,
+						Status:  Retry,
+					},
+					)
+				} else {
+					statuses = append(statuses, BulkSubscribeResponseEntry{
+						EntryId: entry.EntryId,
+						Status:  Drop,
+					},
+					)
+				}
+			}
+
+			resp := BulkSubscribeResponse{
+				Statuses: statuses,
+			}
+			if err != nil {
+				http.Error(w, err.Error(), PubSubHandlerDropStatusCode)
+				return
+			}
+			w.Header().Add("Content-Type", "application/json")
+			w.WriteHeader(http.StatusOK)
+
+			writeBulkStatus(w, resp)
+		})))
+
+	return nil
+}
+
 func writeStatus(w http.ResponseWriter, s string) {
 	status := &common.SubscriptionResponse{Status: s}
 	if err := json.NewEncoder(w).Encode(status); err != nil {
 		http.Error(w, err.Error(), PubSubHandlerRetryStatusCode)
 	}
 }
+
+func writeBulkStatus(w http.ResponseWriter, s BulkSubscribeResponse) {
```

Comment:
Actually just found that it would mess the function a bit with some if else conditions since writeStatus takes a single status whereas bulk takes many so there is difference in the data type of argument